### PR TITLE
Bug 1707905: Update route when ingress domain changes

### PR DESCRIPTION
**Pull request by mrogers950**
Pushing what I have initially while looking into adding tests.

### CODE REVIEW

File: pkg/operator2/route.go [hunk 1]
```diff
@@ -15,7 +16,21 @@ import (
...
+			// we're only interested in the targetName. ingress
+			if !strings.HasPrefix(ig.Host, targetName) {
+				continue
+			}
+			if !strings.HasSuffix(ig.Host, "." + ingress.Spec.Domain) {
+				// the domain has changed, so update the route
+				updateRoute = true
+			}
+		}
+	}
```
Comment (sttts) [inline pkg/operator2/route.go]: this is dangerous style to have if clauses without else ligned up. If you fall through with a modified err you get some funny behaviour.

File: pkg/operator2/route.go [hunk 2]
```diff
@@ -15,7 +16,21 @@ import (
 )
 
 func (c *authOperator) handleRoute(ingress *configv1.Ingress) (*routev1.Route, *corev1.Secret, error) {
+	var updateRoute bool
 	route, err := c.route.Get(targetName, metav1.GetOptions{})
+	if err == nil {
+		// route exists, check to see if the ingress domain has changed
+		for _, ig := range route.Status.Ingress {
+			// we're only interested in the targetName. ingress
+			if !strings.HasPrefix(ig.Host, targetName) {
```
Comment (sttts) [inline pkg/operator2/route.go]: should this be `targetName+ "."`?

File: pkg/operator2/route.go [hunk 3]
```diff
@@ -15,7 +16,21 @@ import (
...
+				continue
+			}
+			if !strings.HasSuffix(ig.Host, "." + ingress.Spec.Domain) {
+				// the domain has changed, so update the route
```
Comment (sttts) [inline pkg/operator2/route.go]: why does the right pre- and postfix imply that the domain has changed?

**Comment by sttts** [comment]
Giving some naive comments. Don't know the code around much.

**Comment by mrogers950** [comment]
@enj I've updated with tests (there may be a few cases I should add). Please check and see if my changes are closer to what you described.

**Comment by enj** [comment]
> @enj I've updated with tests (there may be a few cases I should add). Please check and see if my changes are closer to what you described.

Was easier for me to respond with PR: https://github.com/mrogers950/cluster-authentication-operator/pull/2

File: pkg/operator2/route.go [hunk 4]
```diff
@@ -22,7 +22,14 @@ func (c *authOperator) handleRoute(ingress *configv1.Ingress) (*routev1.Route, *
 	if err != nil {
 		return nil, nil, "FailedCreate", err
 	}
+	defRoute := defaultRoute(ingress)
 
+	if route.Spec.Host != defRoute.Spec.Host {
+		route, err = c.route.Update(defaultRoute(ingress))
```
Comment (enj) [inline pkg/operator2/route.go]: Note that this approach to update will fail because resource version will be unset.

File: pkg/operator2/route.go [hunk 5]
```diff
@@ -36,14 +43,13 @@ func (c *authOperator) handleRoute(ingress *configv1.Ingress) (*routev1.Route, *
...
 	if err := isValidRoute(route, ingress); err != nil {
-		// TODO remove this delete so that we do not lose the early creation timestamp of our route
-		// delete the route so that it is replaced with the proper one in next reconcile loop
-		klog.Infof("deleting invalid route: %#v", route)
-		opts := &metav1.DeleteOptions{Preconditions: &metav1.Preconditions{UID: &route.UID}}
-		if err := c.route.Delete(route.Name, opts); err != nil && !errors.IsNotFound(err) {
-			klog.Infof("failed to delete invalid route: %v", err)
+		klog.Infof("route invalid: %v, restoring", err)
+		routeUpdate := defaultRoute(ingress)
+		routeUpdate.ObjectMeta = route.ObjectMeta // Preserve metadata
```
Comment (enj) [inline pkg/operator2/route.go]: I went with partial preservel in https://github.com/mrogers950/cluster-authentication-operator/pull/2

**Pull request merged**